Implement support for Arrays

NullVoxPopuli opened the pull request:
Repro of some weird behavior: https://runkit.com/nullvoxpopuli/validated-changeset-with-deep-arrays

![image](https://user-images.githubusercontent.com/199018/107816210-6bdf5680-6d42-11eb-8faa-581a423b959d.png)


I have a complex object with objects as array entries that I'd like to manage with a changeset. :D 

the example I have at work is:

```ts
{
  // big object
  listOfStuff: [
     {
       ... stuff
     },
     {
        ... some stuff
     } 
   ]
}
```

I suppose creating arrays when the property is undefined doesn't really make sense, because you can't know if the user does indeed want an array, or a if they want an object with numbered keys that would behave kinda like an array.

I propose that if a value is an array, we set at indicies, rather than properties on an object

NullVoxPopuli commented:
Not sure how you feel about the approach I took here, but the gist is that merging arrays by index is weird -- we need to convert the arrays to objects with index-keys, making the objects "ArrayLike" so that we can merge two objects, rather than two arrays, which we already know how to do. I added [some utilities in a new file](https://github.com/validated-changeset/validated-changeset/pull/104/files#diff-bd1ef4eb51ab1a6a4cb2e9cc832fe5f9bd737e7f24d559a57f58271aeba8801a) to help out with this.

Some outstanding questions:
 - should `changeset.get('some.path.to.array')` always return an array after setting items on the array? [example test](https://github.com/validated-changeset/validated-changeset/pull/104/files#diff-ec61822d929089017a892b4b047229377c9241f16b43252ab851ae39131cdf40R798)
  - should the representation of the set of changes be changed to an array to always match the (sub)content type?

snewcomer commented on test/index.test.ts:
```diff
@@ -771,6 +771,48 @@ describe('Unit | Utility | changeset', () => {
+    expect(dummyChangeset.get('contact.emails')).toEqual(['bob@email.com', 'the_bob@email.com']);
+
+    dummyChangeset.set('contact.emails.0', 'fred@email.com');
+    dummyChangeset.set('contact.emails.1', 'the_fred@email.com');
```
Could we have an assertion where `get/set` is out of range?

NullVoxPopuli replied:
on the lower bound, out of range is only negative, I can add tests for that, but on the upper range, there is no out of bounds, because this is how you'd add stuff.
![image](https://user-images.githubusercontent.com/199018/108102768-fda6d680-7056-11eb-8603-403db96e0346.png)

snewcomer replied:
Oh right.  But do you think it is worthwhile adding a test for this custom implementation when contact.emails is only len 2 but we do the following?

`dummyChangeset.set('contact.emails.10', 'the_fred@email.com');`

NullVoxPopuli replied:
yup! I'm working on adding more tests now (as soon as I figure out how to make `changeset.get('some array').toEqual(some array)`

NullVoxPopuli replied:
TIL 
![image](https://user-images.githubusercontent.com/199018/108126984-8d10b180-7078-11eb-977a-334655b3dd61.png)

I think this should probs be a hard error? even though behavior is kinda undefined in native arrays?

snewcomer commented on test/index.test.ts:
```diff
@@ -771,6 +771,48 @@ describe('Unit | Utility | changeset', () => {
+
+    dummyChangeset.execute();
+    expect(dummyModel.contact.emails).toEqual(['fred@email.com', 'the_fred@email.com']);
+  });
```
Could we also have some assertions for...

- `changeset.changes`
- `changeset.isValid`
- `changeset.save`

NullVoxPopuli replied:
added

snewcomer reviewed (commented):
Thanks for this PR!  I think I am generally in favor of this PR.  A few more assertions to battle test this would be helpful.

Note we also have some preliminary checks before merging you can run locally - `npm run test:ember-changeset && npm run test:ember-changeset-validations`

NullVoxPopuli commented on src/utils/set-deep.ts:
```diff
@@ -60,10 +61,14 @@ export default function setDeep(
+    if (!isComplex) {
       options.safeSet(target, prop, {});
-    } else if (isObj && isChange(target[prop])) {
+    } else if (isComplex && isChange(target[prop])) {
```
>  Are you indicating target[prop] might be an array?

yup, the object-array thing of today works with array's index access, which is cool. _thanks implicit type casting_! (for reals tho, this is one of the few times where I like this part of JS lol)

NullVoxPopuli commented:
Also added:
 - library tests to CI
 - `yarn test:debug:named "part of test name here"` (or I suppose npm run, if that's what you're in to :upside_down_face: ) -- I can never remember all the args needed to debug with jest

NullVoxPopuli commented on src/utils/object-tree-node.ts:
```diff
@@ -11,13 +12,24 @@ const objectProxyHandler = {
    */
   get(node: ProxyHandler, key: string): any {
     if (typeof key === 'symbol') {
+      if (key === Symbol.iterator) {
```
this feels fishy -- all of this is to make jest's expect work... which I feel like is kinda lying -- cause if you console.log(changeset.get('some array'), you get the ObjectNodeProxy, which.. I think is fine and it makes sense that you'd have to call unwrap on it to get the human-readable value out?

NullVoxPopuli commented on test/index.test.ts:
```diff
@@ -771,6 +771,118 @@ describe('Unit | Utility | changeset', () => {
+          { key: 'contact.emails.3', value: 'greg@email.com' }
+        ]);
+
+        expect(changeset.get('contact.emails').unwrap()).toEqual([
```
unwrap required here. 

Is `get` intended to automatically unwrap always?

NullVoxPopuli replied:
legit, thanks!

NullVoxPopuli commented on test/index.test.ts:
```diff
@@ -771,6 +771,182 @@ describe('Unit | Utility | changeset', () => {
+
+        expect(changeset.get('contact.emails.0')).toEqual('fred@email.com');
+        expect(changeset.changes).toEqual([{ key: 'contact.emails.0', value: 'fred@email.com' }]);
+        expect(changeset.get('contact.emails')).toEqual(['fred@email.com']);
```
without the weird jest hacks, this would require .unwrap() to be equallable with an array.
I think the main issue comes from the original object the proxy wraps being an object -- I tried to find a way to change that with Proxy, but it seems the ability to change how an object is iterated over or perceived is deliberately not part of the Proxy spec -- which kinda makes sense, cause it'd probs be super confusing

NullVoxPopuli replied:
it would! _most excellent_

NullVoxPopuli replied:
and yeah, I don't know of any situations where in-app, I'd want to access a non-leaf

snewcomer commented on test/index.test.ts:
```diff
@@ -771,6 +771,182 @@ describe('Unit | Utility | changeset', () => {
+      it('can add properties to an entry', () => {
+        const changeset = Changeset(initialData);
+
+        debugger;
```
☝️

NullVoxPopuli replied:
gasp! thanks!

snewcomer commented on src/utils/object-tree-node.ts:
```diff
@@ -118,11 +131,19 @@ class ObjectTreeNode implements ProxyHandler {
       if (isObject(content)) {
         changes = normalizeObject(changes, this.isObject);
         return { ...content, ...changes };
+      } else if (Array.isArray(content)) {
```
This seems safe for ember-data.  B/c we aren't using the Ember provided `isArray`.

snewcomer reviewed (commented):
This is looking pretty sound!

NullVoxPopuli commented on src/utils/merge-deep.ts:
```diff
@@ -151,9 +152,14 @@ export default function mergeDeep(
     };
   let sourceIsArray = Array.isArray(source);
   let targetIsArray = Array.isArray(target);
+  let sourceIsArrayLike = isArrayObject(source);
```
yup!

NullVoxPopuli commented on test/index.test.ts:
```diff
@@ -771,6 +771,211 @@ describe('Unit | Utility | changeset', () => {
+
+        changeset.set('contact.emails.1', null);
+
+        expect(changeset.get('contact.emails').unwrap()).toEqual([null, null]);
```
idk if this is technically "deleting", but it made me wonder if changeset needs some post-processing callback to run before "save".

cause for some uses, you may want to collapse all the falsey values in an array, like, 
```ts
data.array = data.array.filter(Boolean);
```
but I could also see some use cases where you'd want the falsey values

snewcomer replied:
I think in this case, the recommended way would be to `.set('contact.emails', [])`?  I'm happy leaving the nulls in there if that is what the `set` did!

NullVoxPopuli commented:
also, should this scenario be supported?
```ts
      it('can add a new object at once, and edit it', () => {
        const changeset = Changeset(initialData);

        changeset.set('contact.emails.1', {
          funEmail: 'fun@email.com',
          primary: 'primary@email.com'
        });

        expect(changeset.get('contact.emails.1.funEmail')).toEqual('fun@email.com');
        expect(changeset.get('contact.emails.1.primary')).toEqual('primary@email.com');
        expect(changeset.get('contact.emails').unwrap()).toEqual([
          { primary: 'bob@email.com' },
          { primary: 'primary@email.com', funEmail: 'fun@email.com' }
        ]);
        expect(changeset.changes).toEqual([
          {
            key: 'contact.emails.1',
            value: { funEmail: 'fun@email.com', primary: 'primary@email.com' }
          }
        ]);

        changeset.set('contact.emails.1.primary', 'primary2@email.com');

        expect(changeset.get('contact.emails.1.primary')).toEqual('primary2@email.com');
      });
```

```
  ● Unit | Utility | changeset › arrays of objects within nested objects › #set › can add a new object at once, and edit it

    expect(received).toEqual(expected) // deep equality

    Expected: "primary2@email.com"
    Received: undefined

       997 |         changeset.set('contact.emails.1.primary', 'primary2@email.com');
       998 | 
    >  999 |         expect(changeset.get('contact.emails.1.primary')).toEqual('primary2@email.com');
           |                                                           ^
      1000 |       });
      1001 |     });
      1002 |   });
```

NullVoxPopuli commented:
looks like yes, based on other tests (`#execute keeys prototype of set object` and `#execute it works with nested keys'`

NullVoxPopuli commented:
@snewcomer We should be good to go :+1:

NullVoxPopuli commented on src/utils/set-deep.ts:
```diff
@@ -79,9 +80,14 @@ export default function setDeep(
+        const nestedKeys =
+          Array.isArray(target) || isArrayObject(target)
+            ? keys.slice(i + 1, keys.length).join('.') // remove first key segment as well as the index
+            : keys.slice(1, keys.length).join('.'); // remove first key segment
```
I don't know if the object case needs to also be based off the index, but for the ArrayLike situation, this made sense

NullVoxPopuli commented on test/utils/set-deep.test.ts:
```diff
@@ -228,6 +228,112 @@ describe('Unit | Utility | set deep', () => {
     });
   });
 
+  it('set on nested properties within an empty object', () => {
```
I had to add this test, because my intuitive feelings about how arrays should work wasn't working.

idk if this is intended, or why it would be intended, but intuitively, I would have thoughts this would have been the case:

```ts
// ...
    expect(value).toEqual({
      top: {
        name: new Change('zoo'),
        foo: new Change({
          other: 'baz'
        })
      }
    });
```
because the `foo` is an entirely new object,

and how this correlates to arrays: 
instead of
```ts
it('sets when the array is initially empty', () => {
  const objA = {
    top: []
  };
  let value = setDeep(objA, 'top.0.name', new Change('zoo'));

  expect(value).toEqual({
    top: [{ name: new Change('zoo') }]
  });

  value = setDeep(value, 'top.0.foo.other', new Change('baz'));

  expect(value).toEqual({
    top: [
      {
        name: new Change('zoo'),
        foo: {
          other: new Change('baz')
        }
      }
    ]
  });
 });
```
maybe: 
```ts
it('sets when the array is initially empty', () => {
  const objA = {
    top: []
  };
  let value = setDeep(objA, 'top.0.name', new Change('zoo'));

  expect(value).toEqual({
    top: [new Change({ name: new Change('zoo') })]
  });

  value = setDeep(value, 'top.0.foo.other', new Change('baz'));

  // the big difference here -- the first entry remains an entire single Change
  expect(value).toEqual({
    top: [
      new Change({
        name: 'zoo',
        foo: {
          other: 'baz'
        }
      })
    ]
  });
});
```

NullVoxPopuli commented on test/utils/set-deep.test.ts:
```diff
@@ -228,6 +228,112 @@ describe('Unit | Utility | set deep', () => {
+    });
+  });
+
+  describe('arrays at various nestings within an object', () => {
```
I've started using `describe` blocks for groups of this stuff so I can debug scoped down sections of the test suite, so:
```bash
yarn test:debug:named "arrays at various nestings within an object"
```
allows me to debug just this subset of tests

snewcomer reviewed (approved):
This is a really really nice PR.  Thank you and let me know when you are comfortable for me to merge and release!

NullVoxPopuli commented:
Yeah, I'm good with merging. I think at this point, I need to get our internal work merged (pending releases of this, ember changeset, and ember validated changeset).

Anything I run in to, I can open bugfix prs for

🎉

NullVoxPopuli commented on test/index.test.ts:
```diff
@@ -771,6 +771,249 @@ describe('Unit | Utility | changeset', () => {
+        expect(changeset.errors).toEqual([
+          { key: 'contact.emails.0', validation: 'Fred is banned', value: 'fred@email.com' }
+        ]);
+      });
```
added!

NullVoxPopuli commented on test/index.test.ts:
```diff
@@ -771,6 +771,249 @@ describe('Unit | Utility | changeset', () => {
+        expect(changeset.changes).toEqual([
+          { key: 'contact.emails.1', value: 'fred@email.com' },
+          { key: 'contact.emails.3', value: 'greg@email.com' }
+        ]);
```
added!

NullVoxPopuli commented:
I found an inconsistency depending on the depth of a change:
```ts
changeset.get('contacts.emails') // => requires unwrap
changeset.get('emails') // does not have unwrap
```

I've found that there is a use case for wanting to access non-leaf data: iterating over the data to display navigation to different forms that would open a form to edit each array entry -- so... 
I don't think we should block this PR, as it's gotten kinda big, _but_, I do worry than forcing `changeset.get` to return a that node proxy for all complex objects _could_ be a breaking change

NullVoxPopuli commented:
Turns out, I'm getting faster at this :D 

We should be good to go again :D 
No breaking change needed, since the added proxy wrap in `get` is specifically for arrays

NullVoxPopuli commented:
I just keep finding issues. O.o

@snewcomer I added another test for what I'm running in to now. Idk if you want to take a stab at it. 

But I wonder if some refactoring might need to happen in order to make this easier? 
I _feel_ like this should be an easy addition due to how similar array and object access is -- but it hasn't been. :thinking: 

We also need a consistent way to get the "current value of the subtree at 'this' path", like.. I was thinking maybe a `preview` method that also takes a property path but _always_ returns the source data merged with the changes known for that path.

My recent issues have come from top level properties having inconsistent need for unwrap. Example:
 - have `emails: []`, can access via `changeset.get('emails');`
  - make change to an item within that array
  - `changeset.get('emails')` now needs `unwrap()` (only because I added an Array.isArray check in `get` to return a proxy) -- but without it, there is no merge of data and changes. It seems only the ObjectTreeNode can give a preview.

NullVoxPopuli commented:
Just found out there is _something_ in this PR that fixes normal object behavior. So maybe it is best to merge it and iterate later? thoughts?

snewcomer commented on test/index.test.ts:
```diff
@@ -771,6 +771,463 @@ describe('Unit | Utility | changeset', () => {
+          primary: 'primary0@email.com'
+        },
+        null
+      ]);
```
Does this test pass if you add these assertions (similar to the test below)?

```
expect(changeset.get('emails.0.fun')).toEqual('fun0@email.com');
expect(changeset.get('emails.0.primary')).toEqual('primary0@email.com');
```

NullVoxPopuli replied:
oh interesting. This is the diff with those:
![image](https://user-images.githubusercontent.com/199018/108858204-e5383e00-75b9-11eb-82bd-21062c56066c.png)

NullVoxPopuli replied:
I think this is a more general bug though, and why my work stuff _can't_ use the published validated changeset. 

_sometimes_ when accessing data via `get`, an object returns.
In _this_ case, the object is just totally incorrect. :thinking:

snewcomer replied:
Ok so same as the one below (was just checking if the `value` was the culprit).

Overall, this is a really great PR and I think we are really close.  I think we can focus on `.get` or `ObjectTreeNode` may need to be modified.  I can certainly jump in and see where the last few improvements need to be made.

NullVoxPopuli replied:
I'll take another stab, and see what happens :D

NullVoxPopuli replied:
fixed it -- wasn't so bad after all :D

snewcomer commented on src/index.ts:
```diff
@@ -39,6 +39,7 @@ import {
   ValidatorAction,
   ValidatorMap
 } from './types';
+import { isArrayObject } from './utils/array-object';
```
✂️ 

Should make sure this is caught by linting in CI

NullVoxPopuli replied:
I can PR tomorrow that adds / updates linting in the CI config tomorrow?

snewcomer commented on src/index.ts:
```diff
@@ -955,12 +958,26 @@ export class BufferedChangeset implements IChangeset {
         } else if (typeof result !== 'undefined') {
           return result;
         }
+        const baseContent = this.safeGet(content, baseKey);
```
What is the value of `result` here for the array case?  I thought one of the if expressions above would have been tripped.  It looks like there might be some minor refactors here.

NullVoxPopuli replied:
`result` is `undefined` here. It happens mostly when working with new array entries or previously deleted entries, iirc

snewcomer commented on src/index.ts:
```diff
@@ -955,12 +958,26 @@ export class BufferedChangeset implements IChangeset {
+          // give back an object that can further retrieve changes and/or content
+          const tree = new ObjectTreeNode(subChanges, baseContent, this.getDeep, this.isObject);
+
+          return tree;
```
For the array case, do we need to return an ObjectTreeNode?  Just a little background just in case there is some intersting information for this discussion...

The main reason for the `ObjectTreeNode` is to ensure we don't lose sibling keys.  So an object with both `changes` and unchanged content are able to be retrieved at any arbitrary level of the object. 

```
{
  changeValue: ...,
  original: ... 
}
```

Now for the array scenario, in one case we have `changes` we want to return.  In the other, we have `content` we want to return.  However, in both cases, I imagine that the type is an array and we don't have to deal with merging sibling keys.

NullVoxPopuli replied:
> For the array case, do we need to return an ObjectTreeNode? 

yeah, cause atm, ObjectTreeNode is the only thing that knows that changes are on arrays and can convert between arrays and array objects.

_however_ this changes the behavior of `changeset.get('prop.someArray')` where it now needs `unwrap`, where it previously did not -- but this is more consistent with the deep object changes.

> . However, in both cases, I imagine that the type is an array and we don't have to deal with merging sibling keys.

🤷 
The overloaded behavior of this method is dizzying to me :(

snewcomer reviewed (commented):
I'm really glad the tests are passing!  Just a few questions/comments before we can get this in!
